// @formcrafts/embed/rollup.config.js
import typescript from '@rollup/plugin-typescript';
import css from "rollup-plugin-import-css";
import esbuild from 'rollup-plugin-esbuild'
import { defineConfig } from 'rollup';

const banner = `/**
 * @formcrafts/embed
 * Copyright (c) 2024-${new Date().getFullYear()}, Nishant Agrawal. (MIT Licensed)
 * https://github.com/formcrafts/embed
 */

/**
 * DO NOT EDIT THIS FILE
 * The code in this file is generated from files in ./src/
 */
`;

export default defineConfig([
  {
    input: 'src/embed.ts',
    output: [{
      file: 'lib/embed.esm.js',
      format: 'esm',
      sourcemap: true,
      banner
    },
    {
      file: 'lib/embed.umd.js',
      format: 'umd',
      name: 'embed',
      sourcemap: true,
      banner
    },
    {
      file: 'lib/embed.min.js',
      format: 'umd',
      name: 'embed',
      sourcemap: true,
      banner,
      plugins: [esbuild({
        banner: banner,
        minify: true,
        sourcemap: true,
        legalComments: 'inline'
      })]
    },
    {
      file: 'lib/embed.cjs',
      format: 'cjs',
      name: 'embed',
      sourcemap: true,
      banner
    }],
    plugins: [
      css({
        minify: true,
      }),
      typescript()
    ]
  }
]);
